java-symbolic-execution/src/main/java/org/sonar/java/se/checks/OptionalGetBeforeIsPresentCheck.java

```diff
@@ -209,52 +211,101 @@ public void visitMethodInvocation(MethodInvocationTree tree) {
-        psEmpty.forEach(context::addTransition);
-        psPresent.forEach(context::addTransition);
+        handleOptionalOrElseMethod(tree);
+      } else if (OPTIONAL_OF.matches(tree)) {
+        handleOptionalOfMethod();
+      } else if (OPTIONAL_OF_NULLABLE.matches(tree)) {
+        handleOptionalOfNullableMethod();
+      }
+    }
+
+    private void handleOptionalOfMethod() {
+      constraintManager.setValueFactory(() -> new OptionalSymbolicValue(programState.peekValue()));
+    }
+
+    private void handleOptionalTestMethods(MethodInvocationTree tree) {
+      constraintManager.setValueFactory(() -> new OptionalTestMethodSymbolicValue(programState.peekValue(), tree.symbol()));
+    }
+
+    private void handleOptionalGetMethod(MethodInvocationTree tree) {
+      if (presenceHasNotBeenChecked(programState.peekValueSymbol())) {
+        SymbolicValue optionalSV = Objects.requireNonNull(programState.peekValue());
+        context.addExceptionalYield(optionalSV, programState, "java.util.NoSuchElementException", check);
+        reportIssue(tree);
+        // continue exploration after reporting, assuming the optional is now present (killing any noise after the initial issue)
+        programState = programState.addConstraint(optionalSV, OptionalConstraint.PRESENT);
+      }
+    }
+
+    private String useIsEmptyPrefix(String owner) {
+      return java11 ? String.format(" or \"!%sisEmpty()\"", owner) : "";
+    }
+
+    private void handleOptionalFilterMethod() {
+      // filter has one parameter, so optional is next item on stack
+      SymbolicValue optionalSV = programState.peekValue(1);
+
+      if (programState.getConstraint(optionalSV, OptionalConstraint.class) == OptionalConstraint.NOT_PRESENT) {
+        // reuse the same optional - filtering a non-present optional is a no-op
+        constraintManager.setValueFactory(() -> optionalSV);
+      } else {
+        constraintManager.setValueFactory(() -> new FilteredOptionalSymbolicValue(optionalSV));
+      }
+    }
+
+    private void handleOptionalOrElseMethod(MethodInvocationTree tree) {
+      ProgramState.Pop pop = programState.unstackValue(2);
+      SymbolicValue orElseValue = pop.values.get(0);
+      SymbolicValue optional = pop.values.get(1);
+      List<ProgramState> psEmpty = optional.setConstraint(pop.state.stackValue(orElseValue), OptionalConstraint.NOT_PRESENT);
+      SymbolicValue symbolicValue;
+      if(optional instanceof OptionalSymbolicValue) {
+        symbolicValue = ((OptionalSymbolicValue) optional).wrappedValue;
+      } else {
+        symbolicValue = constraintManager.createSymbolicValue(tree);
+      }
+      List<ProgramState> psPresent = optional.setConstraint(pop.state.stackValue(symbolicValue), OptionalConstraint.PRESENT);
+      psEmpty.forEach(context::addTransition);
+      psPresent.forEach(context::addTransition);
+      programState = null;
```

Comment:
What about adding the '// interrupt current path to only use transitions' comment before this line?